Strenghten preconditions of linalg.cross

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #83798

This makes `linalg.cross` array API complaint (https://github.com/data-apis/array-api/issues/415) and fixes a few bugs.

Fixes https://github.com/pytorch/pytorch/issues/77629
Fixes https://github.com/pytorch/pytorch/issues/83756

lezcano committed:
Strenghten preconditions of linalg.cross

This makes `linalg.cross` array API complaint (https://github.com/data-apis/array-api/issues/415) and fixes a few bugs.

Fixes https://github.com/pytorch/pytorch/issues/77629
Fixes https://github.com/pytorch/pytorch/issues/83756

[ghstack-poisoned]

Lezcano commented:
@kgryte @asmeurer PTAL

lezcano committed:
Update on "Strenghten preconditions of linalg.cross"


This makes `linalg.cross` array API complaint (https://github.com/data-apis/array-api/issues/415) and fixes a few bugs.

Fixes https://github.com/pytorch/pytorch/issues/77629
Fixes https://github.com/pytorch/pytorch/issues/83756

[ghstack-poisoned]

lezcano committed:
Update on "Strenghten preconditions of linalg.cross"


This makes `linalg.cross` array API complaint (https://github.com/data-apis/array-api/issues/415) and fixes a few bugs.

Fixes https://github.com/pytorch/pytorch/issues/77629
Fixes https://github.com/pytorch/pytorch/issues/83756

[ghstack-poisoned]

lezcano committed:
Update on "Strenghten preconditions of linalg.cross"


This makes `linalg.cross` array API complaint (https://github.com/data-apis/array-api/issues/415) and fixes a few bugs.

Fixes https://github.com/pytorch/pytorch/issues/77629
Fixes https://github.com/pytorch/pytorch/issues/83756

[ghstack-poisoned]

lezcano committed:
Update on "Strenghten preconditions of linalg.cross"


This makes `linalg.cross` array API complaint (https://github.com/data-apis/array-api/issues/415) and fixes a few bugs.

Fixes https://github.com/pytorch/pytorch/issues/77629
Fixes https://github.com/pytorch/pytorch/issues/83756

[ghstack-poisoned]

lezcano committed:
Update on "Strenghten preconditions of linalg.cross"


This makes `linalg.cross` array API complaint (https://github.com/data-apis/array-api/issues/415) and fixes a few bugs.

Fixes https://github.com/pytorch/pytorch/issues/77629
Fixes https://github.com/pytorch/pytorch/issues/83756

[ghstack-poisoned]

lezcano committed:
Update on "Strenghten preconditions of linalg.cross"


This makes `linalg.cross` array API complaint (https://github.com/data-apis/array-api/issues/415) and fixes a few bugs.

Fixes https://github.com/pytorch/pytorch/issues/77629
Fixes https://github.com/pytorch/pytorch/issues/83756

[ghstack-poisoned]

Lezcano commented:
All the removed tests from `test_linalg.py` are now tested in the OpInfo.

Lezcano commented on functorch/test/test_vmap.py:
```diff
@@ -3222,6 +3222,7 @@ def test():
         xfail('eye', ''),  # non-tensor input
         xfail('broadcast_shapes', ''),  # test runner can't handle non-Tensor ops
         xfail('sparse.sampled_addmm'),  # sparse
+        xfail('cross'),  # The default value of dim in op is *very* weird. No wonder it doesn't work
```
for `linalg.cross`, the default is -1. For `cross`... well... it actually has one of the craziest defaults in the whole PyTorch.
> If dim is not given, it defaults to the first dimension found with the size 3. Note that this might be unexpected.

Lezcano commented on test/test_linalg.py:
```diff
@@ -4504,48 +4504,6 @@ def test_linalg_cross(self, device, dtype):
         torch.linalg.cross(x, y, dim=1, out=res2)
         self.assertEqual(res1, res2)
 
-        # non contiguous case 1
```
We have non-contiguity tests in OpInfos. I just removed these because they are redundant now. Everything works as expected though.

Lezcano commented:
Addressed the reviews. PTAL @samdow

lezcano committed:
Update on "Strenghten preconditions of linalg.cross"


This makes `linalg.cross` array API complaint (https://github.com/data-apis/array-api/issues/415) and fixes a few bugs.

Fixes https://github.com/pytorch/pytorch/issues/77629
Fixes https://github.com/pytorch/pytorch/issues/83756

[ghstack-poisoned]

lezcano committed:
Update on "Strenghten preconditions of linalg.cross"


This makes `linalg.cross` array API complaint (https://github.com/data-apis/array-api/issues/415) and fixes a few bugs.

Fixes https://github.com/pytorch/pytorch/issues/77629
Fixes https://github.com/pytorch/pytorch/issues/83756

[ghstack-poisoned]

mruberry reviewed (approved):
Thanks @Lezcano!

Lezcano commented:
@pytorchbot merge

janeyx99 commented:
@pytorchbot revert -m "Sorry, land race caused functorch issues https://hud.pytorch.org/pytorch/pytorch/commit/7f0198e7390eff2f2f5fcb33ce36c99ec3b7f55e" -c landrace

samdow commented:
Ahh okay this is coming from the error inputs. I'll file an issue for it so feel free to xfail here

lezcano committed:
Update on "Strenghten preconditions of linalg.cross"


This makes `linalg.cross` array API complaint (https://github.com/data-apis/array-api/issues/415) and fixes a few bugs.

Fixes https://github.com/pytorch/pytorch/issues/77629
Fixes https://github.com/pytorch/pytorch/issues/83756

[ghstack-poisoned]

Lezcano commented:
@pytorchbot merge -g

Second try!

Lezcano commented:
@pytorchbot merge -f "unrelated failure"
